# Esame Finale

**Il committente richiede di avere un report su casi e vaccinazioni in diverse aree del mondo; a tal fine, richiede di utilizzare il dataset, curato da Our World in Data, all'indirizzo https://github.com/owid/covid-19-data/tree/master/public/data; alla voce "🗂 Download our complete COVID-19 dataset" scaricare il dataset in formato CSV oppure in formato JSON.**

In [1]:
import os
import pandas as pd

In [2]:
current_work_directory = os.getcwd()

In [3]:
covid_filepath = os.path.join(current_work_directory,
                             "owid-covid-data.csv")

In [4]:
with open(covid_filepath) as c:
    df_covid = pd.read_csv(c,low_memory=False)
    
df_covid

iso_code continent     location        date  total_cases  new_cases  \
0           AFG      Asia  Afghanistan  2020-01-05          NaN        0.0   
1           AFG      Asia  Afghanistan  2020-01-06          NaN        0.0   
2           AFG      Asia  Afghanistan  2020-01-07          NaN        0.0   
3           AFG      Asia  Afghanistan  2020-01-08          NaN        0.0   
4           AFG      Asia  Afghanistan  2020-01-09          NaN        0.0   
...         ...       ...          ...         ...          ...        ...   
377002      ZWE    Africa     Zimbabwe  2024-01-31     266265.0        0.0   
377003      ZWE    Africa     Zimbabwe  2024-02-01     266265.0        0.0   
377004      ZWE    Africa     Zimbabwe  2024-02-02     266265.0        0.0   
377005      ZWE    Africa     Zimbabwe  2024-02-03     266265.0        0.0   
377006      ZWE    Africa     Zimbabwe  2024-02-04     266265.0        0.0   

        new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
0                      NaN           NaN         0.0                  NaN   
1                      NaN           NaN         0.0                  NaN   
2                      NaN           NaN         0.0                  NaN   
3                      NaN           NaN         0.0                  NaN   
4                      NaN           NaN         0.0                  NaN   
...                    ...           ...         ...                  ...   
377002               4.429        5737.0         0.0                  0.0   
377003               4.429        5737.0         0.0                  0.0   
377004               4.429        5737.0         0.0                  0.0   
377005               4.429        5737.0         0.0                  0.0   
377006               0.000        5737.0         0.0                  0.0   

        ...  male_smokers  handwashing_facilities  hospital_beds_per_thousand  \
0       ...           NaN                  37.746                         0.5   
1       ...           NaN                  37.746                         0.5   
2       ...           NaN                  37.746                         0.5   
3       ...           NaN                  37.746                         0.5   
4       ...           NaN                  37.746                         0.5   
...     ...           ...                     ...                         ...   
377002  ...          30.7                  36.791                         1.7   
377003  ...          30.7                  36.791                         1.7   
377004  ...          30.7                  36.791                         1.7   
377005  ...          30.7                  36.791                         1.7   
377006  ...          30.7                  36.791                         1.7   

        life_expectancy  human_development_index  population  \
0                 64.83                    0.511  41128772.0   
1                 64.83                    0.511  41128772.0   
2                 64.83                    0.511  41128772.0   
3                 64.83                    0.511  41128772.0   
4                 64.83                    0.511  41128772.0   
...                 ...                      ...         ...   
377002            61.49                    0.571  16320539.0   
377003            61.49                    0.571  16320539.0   
377004            61.49                    0.571  16320539.0   
377005            61.49                    0.571  16320539.0   
377006            61.49                    0.571  16320539.0   

        excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                        NaN                          NaN   
1                                        NaN                          NaN   
2                                        NaN                          NaN   
3                                        NaN                          NaN   
4                 

**Punto1: Si richiede di verificare le dimensioni del dataset e le diciture presenti nell'intestazione**

In [12]:
df_covid.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 377007 entries, 0 to 377006
Data columns (total 67 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   iso_code                                    377007 non-null  object 
 1   continent                                   358941 non-null  object 
 2   location                                    377007 non-null  object 
 3   date                                        377007 non-null  object 
 4   total_cases                                 338085 non-null  float64
 5   new_cases                                   366116 non-null  float64
 6   new_cases_smoothed                          364886 non-null  float64
 7   total_deaths                                316113 non-null  float64
 8   new_deaths                                  366124 non-null  float64
 9   new_deaths_smoothed                         364894 non-null  float64
 

In [23]:
df_covid.columns

Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'new_cases_smoothed', 'total_deaths', 'new_deaths',
       'new_deaths_smoothed', 'total_cases_per_million',
       'new_cases_per_million', 'new_cases_smoothed_per_million',
       'total_deaths_per_million', 'new_deaths_per_million',
       'new_deaths_smoothed_per_million', 'reproduction_rate', 'icu_patients',
       'icu_patients_per_million', 'hosp_patients',
       'hosp_patients_per_million', 'weekly_icu_admissions',
       'weekly_icu_admissions_per_million', 'weekly_hosp_admissions',
       'weekly_hosp_admissions_per_million', 'total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'positive_rate', 'tests_per_case', 'tests_units', 'total_vaccinations',
       'people_vaccinated', 'people_fully_vaccinated', 'total_boosters',
       'new_vaccinations', 'new_vaccinations_smoothed',
       't

In [26]:
df_covid = df_covid.loc[:, [
            "iso_code",
            "continent",
            "location",
            "date",
            "total_cases",
            "total_vaccinations"
            ]]

df_covid

iso_code continent     location        date  total_cases  \
0           AFG      Asia  Afghanistan  2020-01-05          NaN   
1           AFG      Asia  Afghanistan  2020-01-06          NaN   
2           AFG      Asia  Afghanistan  2020-01-07          NaN   
3           AFG      Asia  Afghanistan  2020-01-08          NaN   
4           AFG      Asia  Afghanistan  2020-01-09          NaN   
...         ...       ...          ...         ...          ...   
377002      ZWE    Africa     Zimbabwe  2024-01-31     266265.0   
377003      ZWE    Africa     Zimbabwe  2024-02-01     266265.0   
377004      ZWE    Africa     Zimbabwe  2024-02-02     266265.0   
377005      ZWE    Africa     Zimbabwe  2024-02-03     266265.0   
377006      ZWE    Africa     Zimbabwe  2024-02-04     266265.0   

        total_vaccinations  
0                      NaN  
1                      NaN  
2                      NaN  
3                      NaN  
4                      NaN  
...                    ...  
377002                 NaN  
377003                 NaN  
377004                 NaN  
377005                 NaN  
377006                 NaN  

[377007 rows x 6 columns]

**Punto2: Si chiede poi per ogni continente di trovare il numero di casi totali avvenuti in quello stesso continente; si chiede di non considerare eventuali locazioni che nel dataset non appartengono ad alcun continente**

In [7]:
df_covid_full_continent = df_covid.dropna(subset=["continent"])
casi_totali = df_covid_full_continent.groupby("continent")["new_cases"].sum()
df_casi_totali = pd.DataFrame(casi_totali)
df_casi_totali = df_casi_totali.rename(columns = {"new_cases":"total_cases"} )
df_casi_totali

total_cases
continent                 
Africa          13136517.0
Asia           301391174.0
Europe         252013370.0
North America  124526003.0
Oceania         14741735.0
South America   68752780.0

**Punto 3: Sempre riguardo i casi di COVID totali, si chiede di sviluppare una funzione che prenda in input il dataset e due nomi di continenti, e che ne confronti i seguenti relativi descrittori statistici: valori minimo e massimo, media, e percentuale rispetto al numero dei casi totali nel mondo (calcolati anche sulle locazioni senza indicazione di continente)**

In [8]:
import humanize

def confronto_casi_continenti(dataset):
    # passaggio 1 : filtra i due dataset per i continenti selezionati, generando due nuovi dataset
    continente1 = input("Qual è il primo continente che vuoi confrontare?: ").capitalize()
    continente2 = input("Qual è il secondo continente che vuoi confrontare?: ").capitalize()
    df_continente1 = dataset[dataset["continent"].str.capitalize() == continente1]
    df_continente2 = dataset[dataset["continent"].str.capitalize() == continente2]
    # passaggio 2: calcola la somma dei casi totali per ogni continente selezionato
    max_casi_per_stato_continente1 = df_continente1.groupby(["location", "continent"])["total_cases"].agg('max')
    casi_totali_continente1 = max_casi_per_stato_continente1.sum()
    max_casi_per_stato_continente2 = df_continente2.groupby(["location", "continent"])["total_cases"].agg('max')
    casi_totali_continente2 = max_casi_per_stato_continente2.sum()
    # passaggio 3: calcola la somma dei casi totali mondiali (includendo anche i casi in cui continent is null)
    casi_totali_ogni_stato = dataset.groupby(["location", "continent"])["total_cases"].agg('max')
    casi_totali_mondo = casi_totali_ogni_stato.sum()
    # passaggio 4: genera i descrittori statistici per continente 1
    valore_min_casi_continente1 = max_casi_per_stato_continente1.min()
    stato_valore_min_casi_continente1 = max_casi_per_stato_continente1.idxmin()[0]
    valore_max_casi_continente1 = max_casi_per_stato_continente1.max()
    stato_valore_max_casi_continente1 = max_casi_per_stato_continente1.idxmax()[0]
    media_casi_continente1 = max_casi_per_stato_continente1.mean()
    media_casi_continente1 = round(media_casi_continente1, 2)
    confronto_casi_mondiale_continente1 = (casi_totali_continente1 / casi_totali_mondo) * 100
    confronto_casi_mondiale_continente1 = round(confronto_casi_mondiale_continente1, 2)
     # passaggio 5: genera i descrittori statistici per continente 2
    valore_min_casi_continente2 = max_casi_per_stato_continente2.min()
    stato_valore_min_casi_continente2 = max_casi_per_stato_continente2.idxmin()[0]
    valore_max_casi_continente2 = max_casi_per_stato_continente2.max()
    stato_valore_max_casi_continente2 = max_casi_per_stato_continente2.idxmax()[0]
    media_casi_continente2 = max_casi_per_stato_continente2.mean()
    media_casi_continente2 = round(media_casi_continente2, 2)
    confronto_casi_mondiale_continente2 = (casi_totali_continente2 / casi_totali_mondo) * 100
    confronto_casi_mondiale_continente2 = round(confronto_casi_mondiale_continente2, 2)
   # passaggio 6: crea un dataframe riepilogativo per continente 1
    df_riepilogo_stat_continente1 = pd.DataFrame({'Continente': [continente1],
                                                  'Valore minimo di casi totali': [humanize.intcomma(valore_min_casi_continente1)],
                                                  'Stato col valore minimo di casi totali': [stato_valore_min_casi_continente1],
                                                  'Valore massimo di casi totali': [humanize.intcomma(valore_max_casi_continente1)],
                                                  'Stato col valore massimo di casi totali': [stato_valore_max_casi_continente1],
                                                  'Media di casi totali': [humanize.intcomma(media_casi_continente1)],
                                                  'Percentuale dei casi totali rispetto al totale mondiale': [confronto_casi_mondiale_continente1]})
    # passaggio 7: crea un dataframe riepilogativo per continente 2
    df_riepilogo_stat_continente2 = pd.DataFrame({'Continente': [continente2],
                                                  'Valore minimo di casi totali': [humanize.intcomma(valore_min_casi_continente2)],
                                                  'Stato col valore minimo di casi totali': [stato_valore_min_casi_continente2],
                                                  'Valore massimo di casi totali': [humanize.intcomma(valore_max_casi_continente2)],
                                                  'Stato col valore massimo di casi totali': [stato_valore_max_casi_continente2],
                                                  'Media di casi totali': [humanize.intcomma(media_casi_continente2)],
                                                  'Percentuale dei casi totali rispetto al totale mondiale': [confronto_casi_mondiale_continente2]})
    
    df_riepilogo = pd.concat([df_riepilogo_stat_continente1, df_riepilogo_stat_continente2])
    
    return df_riepilogo

In [27]:
confronto_casi_continenti(df_covid)

Qual è il primo continente che vuoi confrontare?:  Oceania
Qual è il secondo continente che vuoi confrontare?:  North America


Continente Valore minimo di casi totali  \
0        Oceania                          4.0   
0  North america                      1,403.0   

  Stato col valore minimo di casi totali Valore massimo di casi totali  \
0                               Pitcairn                  11,769,858.0   
0                             Montserrat                 103,436,829.0   

  Stato col valore massimo di casi totali Media di casi totali  \
0                               Australia           614,237.75   
0                           United States          3,037,218.8   

   Percentuale dei casi totali rispetto al totale mondiale  
0                                               1.90        
0                                              16.08

**Punto4: Si chiede poi di effettuare lo stesso tipo di analisi – anche in questo caso sviluppando una funzione ad hoc – per il numero di vaccinazioni totali per ogni continente**

In [28]:
import humanize

def confronto_vac_continenti(dataset):
    # passaggio 1 : filtra i due dataset per i continenti selezionati, generando due nuovi dataset
    continente1 = input("Qual è il primo continente che vuoi confrontare?: ").capitalize()
    continente2 = input("Qual è il secondo continente che vuoi confrontare?: ").capitalize()
    df_continente1 = dataset[dataset["continent"].str.capitalize() == continente1]
    df_continente2 = dataset[dataset["continent"].str.capitalize() == continente2]
    # passaggio 2: calcola la somma delle vaccinazioni totali per ogni continente selezionato
    vaccinazioni_max_per_stato_continente1 = df_continente1.groupby(["location", "continent"])["total_vaccinations"].agg('max')
    totale_vaccinazioni_continente1 = vaccinazioni_max_per_stato_continente1.sum()
    vaccinazioni_max_per_stato_continente2 = df_continente2.groupby(["location", "continent"])["total_vaccinations"].agg('max')
    totale_vaccinazioni_continente2 = vaccinazioni_max_per_stato_continente2.sum()
    # passaggio 3: calcola la somma delle vaccinazioni totali mondiali (includendo anche i casi in cui continent is null)
    vaccinazioni_totali_ogni_stato = dataset.groupby(["location", "continent"])["total_vaccinations"].agg('max')
    vaccinazioni_totali_mondo =  vaccinazioni_totali_ogni_stato.sum()
    # passaggio 4: genera i descrittori statistici per continente 1
    valore_min_vaccinazioni_continente1 = vaccinazioni_max_per_stato_continente1.min()
    stato_valore_min_vaccinazioni_continente1 = vaccinazioni_max_per_stato_continente1.idxmin()[0]
    valore_max_vaccinazioni_continente1 = vaccinazioni_max_per_stato_continente1.max()
    stato_valore_max_vaccinazioni_continente1 = vaccinazioni_max_per_stato_continente1.idxmax()[0]
    media_vaccinazioni_continente1 = vaccinazioni_max_per_stato_continente1.mean()
    media_vaccinazioni_continente1 = round(media_vaccinazioni_continente1, 2)
    confronto_vaccinazioni_mondiale_continente1 = (totale_vaccinazioni_continente1 / vaccinazioni_totali_mondo) * 100
    confronto_vaccinazioni_mondiale_continente1 = round(confronto_vaccinazioni_mondiale_continente1, 2)
    # passaggio 5: genera i descrittori statistici per continente 2
    valore_min_vaccinazioni_continente2 = vaccinazioni_max_per_stato_continente2.min()
    stato_valore_min_vaccinazioni_continente2 = vaccinazioni_max_per_stato_continente2.idxmin()[0]
    valore_max_vaccinazioni_continente2 = vaccinazioni_max_per_stato_continente2.max()
    stato_valore_max_vaccinazioni_continente2 = vaccinazioni_max_per_stato_continente2.idxmax()[0]
    media_vaccinazioni_continente2 = vaccinazioni_max_per_stato_continente2.mean()
    media_vaccinazioni_continente2 = round(media_vaccinazioni_continente2, 2)
    confronto_vaccinazioni_mondiale_continente2 = (totale_vaccinazioni_continente2 / vaccinazioni_totali_mondo) * 100
    confronto_vaccinazioni_mondiale_continente2 = round(confronto_vaccinazioni_mondiale_continente2, 2)
    # passaggio 6: crea un dataframe riepilogativo per continente 1
    df_riepilogo_stat_continente1 = pd.DataFrame({'Continente': [continente1],
                                                  'Valore minimo di vaccinazioni': [humanize.intcomma(valore_min_vaccinazioni_continente1)],
                                                  'Stato col valore minimo di vaccinazioni': [stato_valore_min_vaccinazioni_continente1],
                                                  'Valore massimo di vaccinazioni': [humanize.intcomma(valore_max_vaccinazioni_continente1)],
                                                  'Stato col valore massimo di vaccinazioni': [stato_valore_max_vaccinazioni_continente1],
                                                  'Media di vaccinazioni': [humanize.intcomma(media_vaccinazioni_continente1)],
                                                  'Percentuale delle vaccinazioni rispetto al totale mondiale': [confronto_vaccinazioni_mondiale_continente1]})
    # passaggio 7: crea un dataframe riepilogativo per continente 2
    df_riepilogo_stat_continente2 = pd.DataFrame({'Continente': [continente2],
                                                  'Valore minimo di vaccinazioni': [humanize.intcomma(valore_min_vaccinazioni_continente2)],
                                                  'Stato col valore minimo di vaccinazioni': [stato_valore_min_vaccinazioni_continente2],
                                                  'Valore massimo di vaccinazioni': [humanize.intcomma(valore_max_vaccinazioni_continente2)],
                                                  'Stato col valore massimo di vaccinazioni': [stato_valore_max_vaccinazioni_continente2],
                                                  'Media di vaccinazioni': [humanize.intcomma(media_vaccinazioni_continente2)],
                                                  'Percentuale delle vaccinazioni rispetto al totale mondiale': [confronto_vaccinazioni_mondiale_continente2]})
    df_riepilogo = pd.concat([df_riepilogo_stat_continente1, df_riepilogo_stat_continente2])
    
    return df_riepilogo

In [29]:
confronto_vac_continenti(df_covid)

Qual è il primo continente che vuoi confrontare?:  Europe
Qual è il secondo continente che vuoi confrontare?:  South America


Continente Valore minimo di vaccinazioni  \
0         Europe                      65,140.0   
0  South america                       4,407.0   

  Stato col valore minimo di vaccinazioni Valore massimo di vaccinazioni  \
0                                  Monaco                  192,221,468.0   
0                        Falkland Islands                  486,436,436.0   

  Stato col valore massimo di vaccinazioni Media di vaccinazioni  \
0                                  Germany         29,095,316.06   
0                                   Brazil         74,197,074.08   

   Percentuale delle vaccinazioni rispetto al totale mondiale  
0                                              11.43           
0                                               7.02

**Punto5: Alla fine, basandosi sui calcoli fatti, il committente chiede di stilare un breve (tre o quattro righe) paragrafo testuale riassuntivo sulle statistiche di casi e vaccinazioni, che si concentri solo sulle differenze esistenti tra Europa, Sud America e Oceania**

Continente | Casi totali MIN | Stato | Casi totali MAX | Stato | Casi totali MEAN | % Globale Casi totali | Vaccinazioni MIN | Stato | Vaccinazioni MAX | Stato | Vaccinazioni MEAN | % Globale Vaccinazioni
-----------|-----------------|------|-----------------|------|----------------|--------------------|-----------------|------|-----------------|------|------------------|------------------------|
Europe|26|Vatican|38,997,490|France|4,941,438.63|32.54|65,140|Monaco|192,221,468|Germany|29,095,316.06|11.43
South America|1,923|Falkland Islands|37,519,960|Brazil|4,910,799.0|8.88|4,407|Falkland Islands|486,436,436|Brazil|74,197,074.08|7.02
Oceania|4|Pitcairn|11,769,858|Australia|614,237.75|1.90|117|Pitcairn|69,306,345|Australia|4,864,102.89|0.64

Confrontando le statistiche sui casi totali e sulle vaccinazioni relative a Europa, Sud America e Oceania emerge come il dato percentuale (sia per i casi totali che per le vaccinazioni) più rilevante sia quello europeo, seguito da quello sudamericano e da quello dell'Oceania. Al tempo stesso si evidenzia come, per tutti e tre i continenti, i valori minimi e massimi per entrambi i fattori siano direttamente proporzionali all'estensione dello Stato e alla sua popolazione totale (per un'analisi più completa sarebbe suggeribile rapportare i valori minimi e massimi alla popolazione dei vari Stati).